File: job_agent/writer.py
from __future__ import annotations

import json
from typing import Any

from job_agent.llm import chat, extract_json
from job_agent.models import CompanyResearch, Critique

DRAFT_SYSTEM = """You are a professional cover-letter writer. Write specific, concrete, non-generic
application materials that reference real facts about the company and map the candidate's
actual experience to the role's requirements. Avoid filler phrases like "I am a hard
worker" — be concrete instead.

Return JSON with this exact shape:
{
  "coverLetter": string,
  "resumeBullets": [string, ...]
}

Rules:
- coverLetter: 250-350 words, reference at least one specific company fact and map 2-3 resume experiences to role requirements
- resumeBullets: 4-6 tailored bullets for this role/company (one line each, strong action verbs)
- Do not invent experience the candidate doesn't have
- Return only the JSON object (no markdown fences, no commentary)"""

CRITIQUE_SYSTEM = """You are a critical reviewer of cover letters. You check for genericness,
unsupported claims, and missed alignment with the job description.

Return JSON with this exact shape:
{
  "hasIssues": boolean,
  "issues": [string, ...],
  "unsupportedClaims": [string, ...],
  "missingAlignment": [string, ...]
}

Be strict but fair. Flag claims not grounded in the resume. Flag missing mappings to
important JD requirements. If the letter is solid, set hasIssues to false and use empty lists."""

REVISE_SYSTEM = """You revise job application materials based on specific critique feedback. Fix only the flagged
issues; keep what already works.

Return JSON with this exact shape:
{
  "coverLetter": string,
  "resumeBullets": [string, ...]
}

Rules:
- Do not invent experience beyond the resume
- Keep cover letter around 250-350 words
- Return only the JSON object (no markdown fences, no commentary)"""


def draft_cover_letter(
    research: CompanyResearch,
    job_description: str,
    resume_text: str,
) -> dict[str, Any]:
    """Writer Step 1: draft tailored application materials."""
    user = f"""Write tailored application materials using the information below.

Company research (structured):
{research.model_dump_json(indent=2)}

Job description:
\"\"\"
{job_description}
\"\"\"

Candidate resume:
\"\"\"
{resume_text}
\"\"\"

Reference at least one specific company fact and map at least 2-3 of the candidate's
real experiences to the role's stated requirements. Do not invent experience the
candidate doesn't have. Return only the JSON object."""

    raw = chat(
        [
            {"role": "system", "content": DRAFT_SYSTEM},
            {"role": "user", "content": user},
        ],
        temperature=0.5,
        json_mode=True,
    )
    return extract_json(raw)


def critique_cover_letter(
    job_description: str,
    resume_text: str,
    draft: dict[str, Any],
) -> Critique:
    """Writer Step 2: self-critique the draft against JD + resume."""
    user = f"""Job description:
\"\"\"
{job_description}
\"\"\"

Candidate resume (ground truth of real experience):
\"\"\"
{resume_text}
\"\"\"

Draft application materials to review:
{json.dumps(draft, indent=2)}

Return only the JSON object."""

    raw = chat(
        [
            {"role": "system", "content": CRITIQUE_SYSTEM},
            {"role": "user", "content": user},
        ],
        temperature=0.1,
        json_mode=True,
    )
    return Critique.model_validate(extract_json(raw))


def revise_cover_letter(
    draft: dict[str, Any],
    critique: Critique,
    resume_text: str,
) -> dict[str, Any]:
    """Writer Step 3: revise only when critique flags issues."""
    user = f"""Original draft:
{json.dumps(draft, indent=2)}

Critique feedback to address:
{json.dumps(critique.model_dump(), indent=2)}

Candidate resume (do not invent anything beyond this):
\"\"\"
{resume_text}
\"\"\"

Rewrite the application materials addressing the feedback. Return only the JSON object."""

    raw = chat(
        [
            {"role": "system", "content": REVISE_SYSTEM},
            {"role": "user", "content": user},
        ],
        temperature=0.4,
        json_mode=True,
    )
    return extract_json(raw)


def write_cover_letter(
    research: CompanyResearch,
    job_description: str,
    resume_text: str,
) -> tuple[dict[str, Any], Critique, dict[str, Any], bool]:
    """Full Writer agent: draft → critique → conditional revise.

    Returns (draft, critique, final_materials, was_revised).
    """
    draft = draft_cover_letter(research, job_description, resume_text)
    critique = critique_cover_letter(job_description, resume_text, draft)

    needs_revision = critique.hasIssues or bool(
        critique.issues or critique.unsupportedClaims or critique.missingAlignment
    )

    if needs_revision:
        final = revise_cover_letter(draft, critique, resume_text)
        return draft, critique, final, True

    return draft, critique, draft, False

File: server/src/agents/writer.js
import { createMessage, extractJson } from "./claude.js";

const DRAFT_SYSTEM = `You are the Writer agent in a multi-agent job application pipeline.

Write specific, grounded application materials using the structured company research and the candidate resume.
Avoid generic filler. Do not invent experience beyond the resume.

Return ONLY JSON (no markdown fences) with this exact shape:
{
  "coverLetter": string,
  "resumeBullets": [string, ...]
}

Rules:
- coverLetter: 250-350 words, references at least one company fact and maps 2-3 resume experiences to role requirements
- resumeBullets: 4-6 tailored bullets for this role/company (each bullet one line, strong action verbs, quantified where resume supports it)`;

const CRITIQUE_SYSTEM = `You critique job application drafts for generic language, unsupported claims, and JD misalignment.

Return ONLY JSON with this exact shape:
{
  "hasIssues": boolean,
  "issues": [string, ...],
  "unsupportedClaims": [string, ...],
  "missingAlignment": [string, ...]
}`;

const REVISE_SYSTEM = `You revise job application materials based on critique feedback.
Fix flagged issues only. Do not invent resume experience.

Return ONLY JSON with this exact shape:
{
  "coverLetter": string,
  "resumeBullets": [string, ...]
}`;

/**
 * @param {Record<string, unknown>} research
 * @param {string} jobDescription
 * @param {string} resumeText
 */
async function draftMaterials(research, jobDescription, resumeText) {
  const response = await createMessage({
    system: DRAFT_SYSTEM,
    messages: [
      {
        role: "user",
        content: `Company research (structured):
${JSON.stringify(research, null, 2)}

Job description:
"""
${jobDescription}
"""

Candidate resume:
"""
${resumeText}
"""`,
      },
    ],
    temperature: 0.5,
  });

  const text = response.content
    .filter((block) => block.type === "text")
    .map((block) => block.text)
    .join("\n");

  return extractJson(text);
}

/**
 * @param {string} jobDescription
 * @param {string} resumeText
 * @param {Record<string, unknown>} draft
 */
async function critiqueMaterials(jobDescription, resumeText, draft) {
  const response = await createMessage({
    system: CRITIQUE_SYSTEM,
    messages: [
      {
        role: "user",
        content: `Job description:
"""
${jobDescription}
"""

Resume (ground truth):
"""
${resumeText}
"""

Draft materials:
${JSON.stringify(draft, null, 2)}`,
      },
    ],
    temperature: 0.1,
  });

  const text = response.content
    .filter((block) => block.type === "text")
    .map((block) => block.text)
    .join("\n");

  return extractJson(text);
}

/**
 * @param {Record<string, unknown>} draft
 * @param {Record<string, unknown>} critique
 * @param {string} resumeText
 */
async function reviseMaterials(draft, critique, resumeText) {
  const response = await createMessage({
    system: REVISE_SYSTEM,
    messages: [
      {
        role: "user",
        content: `Original draft:
${JSON.stringify(draft, null, 2)}

Critique:
${JSON.stringify(critique, null, 2)}

Resume (do not invent beyond this):
"""
${resumeText}
"""`,
      },
    ],
    temperature: 0.4,
  });

  const text = response.content
    .filter((block) => block.type === "text")
    .map((block) => block.text)
    .join("\n");

  return extractJson(text);
}

/**
 * @param {Record<string, unknown>} research
 * @param {string} jobDescription
 * @param {string} resumeText
 * @param {(event: import('./pipeline.js').TraceEvent) => void} emit
 */
export async function runWriter(research, jobDescription, resumeText, emit) {
  emit({
    type: "agent_start",
    agent: "writer",
    message: "Drafting cover letter and tailored resume bullets",
    timestamp: new Date().toISOString(),
  });

  const draft = await draftMaterials(research, jobDescription, resumeText);

  emit({
    type: "draft",
    agent: "writer",
    message: "Initial draft complete",
    payload: draft,
    timestamp: new Date().toISOString(),
  });

  const critique = await critiqueMaterials(jobDescription, resumeText, draft);

  emit({
    type: "critique",
    agent: "writer",
    message: critique.hasIssues ? "Critique flagged issues" : "Draft passed critique",
    payload: critique,
    timestamp: new Date().toISOString(),
  });

  const needsRevision =
    critique.hasIssues ||
    (Array.isArray(critique.issues) && critique.issues.length > 0) ||
    (Array.isArray(critique.unsupportedClaims) && critique.unsupportedClaims.length > 0) ||
    (Array.isArray(critique.missingAlignment) && critique.missingAlignment.length > 0);

  let finalMaterials = draft;
  let revised = false;

  if (needsRevision) {
    emit({
      type: "revise",
      agent: "writer",
      message: "Revising materials based on critique",
      timestamp: new Date().toISOString(),
    });
    finalMaterials = await reviseMaterials(draft, critique, resumeText);
    revised = true;
  }

  emit({
    type: "agent_complete",
    agent: "writer",
    message: revised ? "Revision complete" : "Using draft as final output",
    payload: finalMaterials,
    timestamp: new Date().toISOString(),
  });

  return { draft, critique, finalMaterials, revised };
}
